src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs

```diff
@@ -24,94 +23,82 @@ public void Dispose()
 
         public void Init()
         {
-            foreach (var m in storageQueue.GetAllMessages())
-            {
-                var messageTypeString = m.Body as string;
-                var messageType = new MessageType(messageTypeString); //this will parse both 2.6 and 3.0 type strings
-                var subscriber = Deserialize(m.Label);
+            var messages = storageQueue.GetAllMessages()
+                .OrderByDescending(m => m.ArrivedTime)
+                .ThenBy(x => x.Id) // ensure same order of messages with same timestamp accross all endpoints
+                .ToArray();
 
-                entries.Add(new Entry
+            lock (lookup)
```

Comment:
Do we really need this lock here? Isn't safe to assume that the bootstrapping phase of the endpoint including the feature startup will call init only once? And if it wouldn't it would call this piece of code twice during bootstrapping which would end up just replacing the existing content of the dictionary.